snapshotter/devmapper/pool_device.go

```diff
@@ -0,0 +1,297 @@
+		log.G(ctx).Debugf("reloading existing pool '%s'", poolPath)
+		if err := devicemapper.ReloadPool(poolName, dataFile, metaFile, blockSizeSectors); err != nil {
+			return nil, errors.Wrapf(err, "failed to reload pool '%s'", poolName)
+		}
+	} else {
+		if !os.IsNotExist(err) {
+			return nil, errors.Wrapf(err, "failed to stat for '%s'", poolPath)
+		}
+
+		log.G(ctx).Debug("creating new pool device")
+		if err := devicemapper.CreatePool(poolName, dataFile, metaFile, blockSizeSectors); err != nil {
+			return nil, errors.Wrapf(err, "failed to create thin-pool with name '%s'", poolName)
+		}
+	}
+
+	return &PoolDevice{
+		poolName: poolName,
+		devices:  make(map[string]int),
+	}, nil
+}
+
+func (p *PoolDevice) CreateThinDevice(deviceName string, virtualSizeBytes uint64) (int, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	if _, ok := p.devices[deviceName]; ok {
+		return 0, errors.Errorf("device with name '%s' already created", deviceName)
+	}
+
+	// Create device, retry if device id is taken
+	deviceID, err := p.tryAcquireDeviceID(func(thinDeviceID int) error {
+		return devicemapper.CreateDevice(p.poolName, thinDeviceID)
+	})
+
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to create thin device")
+	}
+
+	p.devices[deviceName] = deviceID
+
+	devicePath := p.GetDevicePath(p.poolName)
+	if err := devicemapper.ActivateDevice(devicePath, deviceName, deviceID, virtualSizeBytes); err != nil {
+		return 0, errors.Wrap(err, "failed to activate thin device")
+	}
+
+	return deviceID, nil
+}
+
+func (p *PoolDevice) CreateSnapshotDevice(deviceName string, snapshotName string, virtualSizeBytes uint64) (int, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	deviceID, ok := p.devices[deviceName]
+	if !ok {
+		return 0, errors.Errorf("device '%s' not found", deviceName)
+	}
+
+	_, ok = p.devices[snapshotName]
+	if ok {
+		return 0, errors.Errorf("snapshot with name '%s' already exists", snapshotName)
+	}
+
+	// Send 'create_snap' message to pool-device
+	devicePoolPath := p.GetDevicePath(p.poolName)
+	thinDevicePath := p.GetDevicePath(deviceName)
+	snapshotDeviceID, err := p.tryAcquireDeviceID(func(snapshotDeviceID int) error {
+		return devicemapper.CreateSnapDevice(devicePoolPath, snapshotDeviceID, thinDevicePath, deviceID)
+	})
+
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to create snapshot")
+	}
+
+	// Activate snapshot
+	if err := devicemapper.ActivateDevice(devicePoolPath, snapshotName, snapshotDeviceID, virtualSizeBytes); err != nil {
+		return 0, errors.Wrap(err, "failed to activate snapshot device")
+	}
+
+	p.devices[snapshotName] = snapshotDeviceID
+	return snapshotDeviceID, nil
+}
+
+func (p *PoolDevice) RemoveDevice(deviceName string) error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	return p.removeDevice(deviceName, true)
+}
+
+func (p *PoolDevice) GetDevicePath(deviceName string) string {
+	return getDevMapperPath(deviceName)
+}
+
+func (p *PoolDevice) Close(ctx context.Context, removeDevices, removePool bool) error {
```

Comment:
How do you feel about separating this into three functions? `Close` closes everything. `CloseDevices` closes devices and `ClosePools` closes pools.

Reply:
I'll review this in subsequent PRs, as it'll going to be refactored.